### PR TITLE
implement full spec for writable streams

### PR DESCRIPTION
**Pull request by bkw**
I had a go at implementing the full spec for writable streams. Is this overkill?

Things done:
- inherit from Stream instead of only EventEmitter
- refactored any changes to _offset into methods _moveOffset and _setOffset
- implemented end() and destroy()
- emit E_PIPE when attempted to write after end()
- emit drain on resume
- added test against stream-spec by @dominictarr

### CODE REVIEW

File: lib/Reader.js [hunk 1]
```diff
@@ -191,5 +187,47 @@ Reader.prototype.compact = function() {
     return;
   }
   this._buffer = this._buffer.slice(this._offset);
+  this._setOffset(0);
```
Comment (bkw) [inline lib/Reader.js:190]: github's diff output is misleading here, this chunk actually applied further down below.

**Comment by felixge** [comment]
This patch LGTM for most parts, but we'll need a good benchmark suite before I'd consider merging. I'm super busy right now, but if you want to do this, I can provide all the information. I'm mostly interested in covering one particular test case that is relevant for node-mysql, but is IMO also fairly representative for other scenarios.

**Comment by bkw** [comment]
If you can supply the information about the benchmarks, I'll give it a try. Wanted to do some speed tuning anyway since I saw your talk on benchmark driven programming.

My guess would be that the overhead for the patch is negligable, but we shall see :-)

Take your time and don't feel rushed, this was just an exercise for me, I don't need the code merged.

**Comment by felixge** [comment]
So, I basically want the same benchmark / parser setup as here:

https://github.com/felixge/faster-than-c/tree/master/figures/mysql2-vs-new-parser

However, this thing misses the (180 MB) fixture that is needed right now. This fixture can also be generated in-memory fairly quickly (and in fact would give me an excuse to work on the `Writer` part I want buffy to have). I'll probably have some time to hack on this during node-dublin, but if you want to go ahead:

http://dev.mysql.com/doc/internals/en/text-protocol.html#packet-ProtocolText::ResultsetRow
http://felixge.s3.amazonaws.com/12/100k-blog-rows.mysql

(The benchmark is basically just parsing through 100k mysql rows containing blog post data, we only need a benchmark for buffy at this point, but comparing to other libraries in the future would be fun).

Anyway, my main problem with the current patch is the `_moveOffset()` method, if you can could re-submit the patch in a way that doesn't include that, I'd merge it right away. Otherwise please let me know if you want to work on the benchmark stuff / wait for me with it.

**Comment by bkw** [comment]
I'll take a look. Unfortunately I really should be doing other things right now. We'll see who's faster :-)

Something like the _moveOffset/_setOffset-Combo is needed for handling end() and destroy(). We need to know whether we have flushed all of our buffer's contents, so we can emit close. I can't see an easy way around this. Maybe defer the checks via nextTick() to after the returning of the value?

But I agree we should get the benchmarks going first. I may have a pleasant surprise for you in the performance department (See #11), but we need real numbers for that.

Have fun in Dublin!

**Comment by felixge** [comment]
Ok, that makes sense, I'll take a closer look very soon! Thank you for your help, and sorry for bottlenecking this change.

**Comment by felixge** [comment]
So, I changed my mind. Please merge this.

I'll do my performance analysis when I find time. For now let's just make this library nice. We can always change stuff again if needed.

**Comment by bkw** [comment]
thx

**Pull request merged**